bark/world/evaluation/ltl/label_functions/safe_distance_label_function.cpp

```diff
@@ -8,77 +8,171 @@
+bool SafeDistanceLabelFunction::IsOncomingVehicle(const bark::world::AgentPtr& front_agent,
+           const bark::world::AgentPtr& rear_agent) const {
+  // Angular deviation allowed around orientation difference of 180
+  using bark::models::dynamic::StateDefinition;
+  const double max_angular_deviation = 5.0 / 360.0 * bark::geometry::B_2PI;
+  const double lower_angular_range = bark::geometry::B_PI - max_angular_deviation;
+  const double upper_angular_range = bark::geometry::B_PI + max_angular_deviation;
+  const double angle_difference = front_agent->GetCurrentState()[StateDefinition::THETA_POSITION] - 
+                                  rear_agent->GetCurrentState()[StateDefinition::THETA_POSITION];
+  const double normed_angle = bark::geometry::Norm0To2PI(angle_difference);
+  const bool is_oncoming = normed_angle < upper_angular_range &&
+         normed_angle > lower_angular_range;
+  VLOG(5) << "oncoming?:" << is_oncoming << ", lar=" << lower_angular_range << ", uar=" 
+          << upper_angular_range << ", na=" << normed_angle;
+  return is_oncoming;
+}
+
+
+bool SafeDistanceLabelFunction::EvaluateEgoCorridor(
     const world::ObservedWorld& observed_world) const {
   auto ego = std::const_pointer_cast<Agent>(observed_world.GetEgoAgent());
   auto lane_corr =
       ego->GetRoadCorridor()->GetNearestLaneCorridor(ego->GetCurrentPosition());
   double frac;
   if (use_frac_param_from_world_) {
-    frac = observed_world.GetFracLateralOffset();
+    frac = observed_world.GetLateralDifferenceThreshold();
   } else {
-    frac = frac_lateral_offset_;
+    frac = lateral_difference_threshold_;
   }
   auto fr_agents = observed_world.GetAgentFrontRearForId(
         ego->GetAgentId(), lane_corr, frac);
   bool distance_safe = true;
+
   if (to_rear_ && fr_agents.rear.first) {
-    distance_safe = CheckSafeDistance(fr_agents.rear.first, ego,
-                                      fr_agents.rear.second, a_o_, a_e_);
-  } else if (fr_agents.front.first) {
-    distance_safe = CheckSafeDistance(ego, fr_agents.front.first,
-                                      fr_agents.front.second, a_e_, a_o_);
+    double v_r = fr_agents.rear.second.to.vlon;
+    double v_f = ego->GetCurrentState()(StateDefinition::VEL_POSITION);
+    double dist = fr_agents.front.second.lon;;
+    distance_safe = CheckSafeDistance(v_f, v_r,
+                                      dist, a_o_, a_e_, delta_others_);
+  } 
+  
+  if (fr_agents.front.first) {
+    if (IsOncomingVehicle(fr_agents.front.first, ego)) {
+      return true;
+    }
+    double v_r = ego->GetCurrentState()(StateDefinition::VEL_POSITION);
+    double v_f = fr_agents.front.second.to.vlon;
+    double dist = fr_agents.front.second.lon;
+    distance_safe = distance_safe && CheckSafeDistance(v_f, v_r,
+                                      dist, a_e_, a_o_, delta_ego_);
   }
-  return {{GetLabel(), distance_safe}};
+  return distance_safe;
+}
+
+bool SafeDistanceLabelFunction::EvaluateCrossingCorridors(
```

Comment:
I want to calculate a safe distance for other vehicles not within the ego vehicles lane corridor. It assumes that the ego is close to the boundary of the other agents corridor to circument the problem of definign a prediction for the ego. When you are close to the boundary you can assume that the ego is within a region soon where the safe distance is meaningful to calculate. 
![IMG_20210325_165406](https://user-images.githubusercontent.com/49236264/112502970-d399b000-8d8a-11eb-8fe5-24e5cd8534b2.jpg)

Reply:
Okay. I guess it would be better to calculate a crossing point, and then calculate the safe distance towards that by calculating longitudinal distances in each vehicle's local frenet system. But if it worked for you, I guess we can leave it like that...